src/app/api/toy-example/README.md

```diff
@@ -0,0 +1,398 @@
+    message: z.literal("The password is invalid"),
+  }),
+
+  z.object({
+    code: z.literal("UNKNOWN"),
+    data: z.any(),
+  }),
+]);
+```
+
+I hope the `SignInRequest` is intuitive. The `SignInResponse` is slightly more complex, but I will explain the motivation. From the specifications, we can determine 4 outcomes when a new user sign up:
+
+1. The user successfully registered (`SUCCESS`),
+2. The email already existed (`EMAIL_EXIST`),
+3. The email is improperly formatted (`INVALID_EMAIL`),
+4. The password is improperly formatted (`INVALID_PASSWORD`).
+
+We added the 5th case (`UNKNOWN`) as a catch-all for unanticipated errors - such as network or database error. Instead of creating 5 different `z.object`, we use [discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) to differentiate between the cases. Discriminated union is a complex TypeScript feature, which we will not do a deep-dive on. At a high level, it enforces type safety. In the context of web development, this means that frontend developers can safely work with the response from the server. Hopefully, the next example will demonstrate how powerful discriminate unions are.
+
+```tsx
+// SignInForm.tsx
+
+import React from "react";
+import { SignInResponse } from "./route.schema";
+
+const SignInForm = () => {
+  const [email, setEmail] = React.useState("");
+  const [password, setPassword] = React.useState("");
+
+  const onSignUp = async () => {
+    // When users signup, make an POST API request to the server
+    const res = await fetch("/api/signup/", {
+      method: "POST",
+      body: JSON.stringify({ email: email, password: password }),
+    });
+    const responseObject = SignInResponse.parse(await res.json());
+
+    // The status code will help TypeScript to determine the type of message
+    if (responseObject.code === "SUCCESS") {
+      // TypeScript tells us that message is "The new user was successfully
+      // registered" when your mouse hovers over res.message
+      responseObject.message;
+    } else if (responseObject.code === "EMAIL_EXIST") {
+      // TypeScript tells us that message is "The email existed already" when your mouse hovers over res.message
+      responseObject.message;
+    }
+  };
+
+  return (
+    <form>
+      <input
+        type="email"
+        value={email}
+        onChange={(e) => setEmail(e.target.value)}
+      />
+      <input
+        type="password"
+        value={password}
+        onChange={(e) => setPassword(e.target.value)}
+      />
+      <input type="submit" value="Submit" onClick={onSignUp} />
+    </form>
+  );
+};
+
+export default SignInForm;
+```
+
+### Why Bother?
+
+You may wonder why are we spending so much time on the specifications?
+
+#### Writing Good Software
+
+Determining the expected input and output helps us think about edge cases and write "delibrated" code - think of your experience in COMP15 and COMP40!
```

Comment:
I would say "carefully considered" instead of "deliberated," it took me a second to wrap my head around what "deliberated" means in this context